shared/lsp/registration.ts

```diff
@@ -389,33 +389,47 @@ export async function register({
         }
     }
 
+    function addRoot(root: sourcegraph.WorkspaceRoot): void {
+        const connectionPromise = (async () => {
+            try {
+                const serverRootUri = clientToServerURI(
+                    new URL(root.uri.toString())
+                )
+                const connection = await connect({
+                    clientRootUri: new URL(root.uri.toString()),
+                    initParams: {
+                        processId: null,
+                        rootUri: serverRootUri.href,
+                        capabilities: clientCapabilities,
+                        workspaceFolders: null,
+                        initializationOptions,
+                    },
+                    registerProviders: true,
+                })
+                subscriptions.add(connection)
+
+                connection.closeEvent.subscribe(() => {
```

Comment:
I may be missing something here, but won't this leak connections?

When a root is removed from the workspace (eg. when navigating away from a file), I would expect the connection to be unsubscribed from: https://sourcegraph.com/github.com/sourcegraph/code-intel-extensions/-/blob/shared/lsp/connection.ts#L86-89

This will call `socket.close()`, which in turn will cause the `close` event to fire on the socket, following which... We'll add another root.

Reply:
Removing a root should first remove it from the map, the. Unsubscribe. That’s why we check if there’s still an entry for it.

Reply:
I see (sorry, I missed "This won't re-open connections that have been closed due to navigation" in the PR description).